src/NServiceBus.Core/Causation/AttachCausationHeadersBehavior.cs

```diff
@@ -9,29 +9,42 @@ class AttachCausationHeadersBehavior : IBehavior<IOutgoingPhysicalMessageContext
     {
         public Task Invoke(IOutgoingPhysicalMessageContext context, Func<IOutgoingPhysicalMessageContext, Task> next)
         {
-            ApplyHeaders(context);
+            IncomingMessage incomingMessage;
+            context.TryGetIncomingPhysicalMessage(out incomingMessage);
+
+            SetRelatedToHeader(context, incomingMessage);
+            SetConversationIdHeader(context, incomingMessage);
 
             return next(context);
         }
 
-        static void ApplyHeaders(IOutgoingPhysicalMessageContext context)
+        static void SetRelatedToHeader(IOutgoingPhysicalMessageContext context, IncomingMessage incomingMessage)
         {
-            var conversationId = CombGuid.Generate().ToString();
+            if (incomingMessage == null)
+            {
+                return;
+            }
 
-            IncomingMessage incomingMessage;
+            context.Headers[Headers.RelatedTo] = incomingMessage.MessageId;
+        }
 
-            if (context.TryGetIncomingPhysicalMessage(out incomingMessage))
+        static void SetConversationIdHeader(IOutgoingPhysicalMessageContext context, IncomingMessage incomingMessage)
+        {
+            string conversationIdFromCurrentMessageContext;
+            if (incomingMessage != null && incomingMessage.Headers.TryGetValue(Headers.ConversationId, out conversationIdFromCurrentMessageContext))
             {
-                context.Headers[Headers.RelatedTo] = incomingMessage.MessageId;
+                context.Headers[Headers.ConversationId] = conversationIdFromCurrentMessageContext;
+                return;
+            }
 
-                string conversationIdFromCurrentMessageContext;
-                if (incomingMessage.Headers.TryGetValue(Headers.ConversationId, out conversationIdFromCurrentMessageContext))
-                {
-                    conversationId = conversationIdFromCurrentMessageContext;
-                }
+            string userDefinedConversationId;
+            if (context.Headers.TryGetValue(Headers.ConversationId, out userDefinedConversationId))
+            {
+                // do not override user defined conversation id if no incoming message exists.
+                return;
```

Comment:
yeah good suggestion. Logging would be an easy change but also easy to miss. Any concerns throwing an exception after a minor version update in case the code was setting this header before?
thoughts @Particular/nservicebus-maintainers ?